compiler/rustc_codegen_llvm/src/back/archive.rs

```diff
@@ -332,22 +402,46 @@ impl<'a> LlvmArchiveBuilder<'a> {
         }
     }
 
-    fn i686_decorated_name(import: &DllImport) -> CString {
+    fn i686_decorated_name(import: &DllImport, mingw: bool) -> String {
         let name = import.name;
+        let prefix = if mingw { "" } else { "_" };
         // We verified during construction that `name` does not contain any NULL characters, so the
         // conversion to CString is guaranteed to succeed.
-        CString::new(match import.calling_convention {
-            DllCallingConvention::C => format!("_{}", name),
-            DllCallingConvention::Stdcall(arg_list_size) => format!("_{}@{}", name, arg_list_size),
+        match import.calling_convention {
+            DllCallingConvention::C => format!("{}{}", prefix, name),
+            DllCallingConvention::Stdcall(arg_list_size) => {
+                format!("{}{}@{}", prefix, name, arg_list_size)
+            }
             DllCallingConvention::Fastcall(arg_list_size) => format!("@{}@{}", name, arg_list_size),
             DllCallingConvention::Vectorcall(arg_list_size) => {
                 format!("{}@@{}", name, arg_list_size)
             }
-        })
-        .unwrap()
+        }
     }
 }
 
 fn string_to_io_error(s: String) -> io::Error {
     io::Error::new(io::ErrorKind::Other, format!("bad archive: {}", s))
 }
+
+fn find_binutils_dlltool(sess: &Session) -> OsString {
+    if sess.host.llvm_target != sess.target.llvm_target {
+        let prefixed_dlltool = if sess.target.arch == "x86" {
+            "i686-w64-mingw32-dlltool"
+        } else {
+            "x86_64-w64-mingw32-dlltool"
+        };
+        let prefixed_dlltool = if cfg!(windows) {
+            [prefixed_dlltool, "exe"].concat()
+        } else {
+            prefixed_dlltool.to_string()
+        };
+        for dir in env::split_paths(&env::var_os("PATH").unwrap_or_default()) {
+            let full_path = dir.join(&prefixed_dlltool);
+            if full_path.is_file() {
+                return full_path.into_os_string();
+            }
+        }
+    }
+    OsString::from("dlltool")
```

Comment:
This is native triple so it should have `.exe` extension.

Reply:
The shell should not have a problem without the `.exe` extension, right?

Reply:
MSYS2 won't have problem but since rustc is native Windows program it doesn't rely on it and instead calls the program directly.

Reply:
This is not fixed yet and shouldn't have been marked as resolved.

Reply:
Sorry, confusion on PR etiquette.

Reply:
Sure, latest commits do should resolve this.